### PR TITLE
[RFC] man.vim: highlight bold and underlined text

### PR DESCRIPTION
**Pull request by keidax**
I was inspired by #5847, but wanted to keep the original syntax highlighting colors:
screenshot 2017-11-22 at 8 39 29 pm
As much as I optimized the Vimscript version it felt a little too slow. Working in Lua speeds things up nicely. My (very informal) testing with LuaJIT shows a roughly 10% overall increase in the time to start `nvim` and render a man page.

It doesn't look like there's any other Lua code included in the runtime yet, but I hope this can still be considered.

I also didn't handle any ANSI SGR sequences, since `man` on my system just uses backspaces, but if there's interest I can try to include those too.

### CODE REVIEW

**Comment by justinmk** [comment]
This is cool! We certainly aren't opposed to using Lua for all runtime code that Nvim project maintains (need to make a list of the exact files somewhere ...). And we definitely forked man.vim so forking it even more is fine :)

**Comment by mhinz** [comment]
Works fine on macOS!

**Comment by mhinz** [comment]
Can this be merged?

**Comment by nhooyr** [comment]
I'd like to review it first.

File: runtime/syntax/man.vim [hunk 1]
```diff
@@ -18,6 +18,18 @@ highlight default link manOptionDesc     Constant
...
 highlight default link manSubHeading     Function
 
+function! s:init_highlight_groups()
+  highlight default manUnderline cterm=underline gui=underline
+  highlight default manBold      cterm=bold      gui=bold
+endfunction
+
+augroup man_init_highlight_groups
+  autocmd!
+  autocmd ColorScheme * call s:init_highlight_groups()
```
Comment (justinmk) [inline runtime/syntax/man.vim]: why is this needed? `highlight default` is a one-time thing.

Comment (keidax) [reply runtime/syntax/man.vim]: I needed it in order to keep the highlighting across colorscheme changes. AFAICT `highlight default link`s are preserved after a `highlight clear`, but groups (other than the builtin ones) with defined default attributes get cleared. Am I missing something?

**Comment by keidax** [comment]
Rebased, added escape sequence handling (for bold, italics, and underlines), plus some test cases (just for highlighting).

File: runtime/syntax/man.vim (hunk 1 above)
Comment (justinmk) [reply runtime/syntax/man.vim]: Does this happen on startup, or only if user changes colorscheme at runtime? The latter is not important. And why wouldn't this also be needed for the other `hi default`'s, e.g. on line 19:

    highlight default link manSubHeading     Function

More generally, syntax needs to be re-initialized if a colorscheme or whatever nukes it. Every plugin cannot be expected to deal with that ad-hoc.

File: runtime/autoload/man.vim [hunk 2]
```diff
@@ -388,4 +387,14 @@ function! man#init_pager() abort
   execute 'silent file man://'.fnameescape(ref)
 endfunction
 
+function! man#highlight_formatted_text() abort
+  let l:modifiable = &modifiable
+  set modifiable
```
Comment (justinmk) [inline runtime/autoload/man.vim]: This could be done in the Lua code via `vim.api.nvim_command()` or whatever. Then we can eliminate this "public" VimL function.

File: runtime/lua/man.lua [hunk 3]
```diff
@@ -0,0 +1,141 @@
+local function highlight_formatted(line, linenr)
+  local chars = {}
+  local prev_char = ''
+  local overstrike, escape = false, false
+  local hls = {} -- Store highlight groups as { attr, start, end }
```
Comment (nhooyr) [inline runtime/lua/man.lua]: This is really janky imo to not name the fields stored in that tuple but I'm not sure if there is a better way. I have very little lua experience.

File: runtime/lua/man.lua [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    end
+  end
+
+  -- Break input into UTF8 characters
+  for char in line:gmatch("[^\128-\191][\128-\191]*") do
+    if overstrike then
+      local last_hl = hls[#hls]
+      if char == prev_char then
+        if char == '_' and attr == UNDERLINE and last_hl and last_hl[3] == byte then
+          -- This underscore is in the middle of an underlined word
```
Comment (nhooyr) [inline runtime/lua/man.lua]: When would this occur? Could you provide some sample input to make the comment more clear?

And also, is there any precedence for this behaviour on when to underline vs bold in regards to an overstriken underscore? To me, it seems intuitive that it would always become bolded.

**Review by nhooyr — CHANGES_REQUESTED** [review]
I found this really hard to read but that may just be because I have very little experience with Lua.

Would appreciate a review from someone else more familiar with Lua.

File: runtime/lua/man.lua [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  end
+
+  local function add_attr_hl(code)
+    local on = true
+    if code == 0 then
+      attr = NONE
+      on = false
+    elseif code == 1 then
+      attr = BOLD
+    elseif code == 21 or code == 22 then
```
[the remark this answers is not in the document]
Comment (keidax) [reply runtime/lua/man.lua]: Good point. I was looking at the [Wikipedia page](https://en.wikipedia.org/wiki/ANSI_escape_code#SGR_.28Select_Graphic_Rendition.29_parameters) which indicates "Bold off or Double Underline". I never saw 21 used anywhere, so I'll drop it.

File: runtime/lua/man.lua [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+          end_attr_hl(a)
+        end
+      else
+        end_attr_hl(attr)
+      end
+    end
+  end
+
+  -- Break input into UTF8 characters
+  for char in line:gmatch("[^\128-\191][\128-\191]*") do
```
[the remark this answers is not in the document]
Comment (keidax) [reply runtime/lua/man.lua]: `"."` will just match a single byte. But the input is encoded as UTF-8, so matching one byte at a time won't work if any characters fall outside of the ASCII range. [More details here.](https://stackoverflow.com/a/15980690/910109)

_Side note: I've only ever seen UTF-8, but I wonder if other encodings would be possible based on terminal/locale settings? If so, would likely need some sort of Lua string library to deal with._

File: runtime/lua/man.lua (hunk 3 above)
Comment (keidax) [reply runtime/lua/man.lua]: Yeah, I'll name them

File: runtime/lua/man.lua [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        hls[#hls + 1] = {attr, byte, byte + #char}
+      end
+
+      overstrike = false
+      prev_char = ''
+      byte = byte + #char
+      chars[#chars + 1] = char
+    elseif escape then
+      -- Use prev_char to store the escape sequence
+      prev_char = prev_char .. char
```
[the remark this answers is not in the document]
Comment (keidax) [reply runtime/lua/man.lua]: Because it's matching against the entire escape sequence, not just the last character.

File: runtime/lua/man.lua [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+      local sgr = prev_char:match("^%[([\020-\063]*)m$")
+      if sgr then
+        local match = ''
+        while sgr and #sgr > 0 do
+          match, sgr = sgr:match("^(%d*);?(.*)")
```
[the remark this answers is not in the document]
Comment (keidax) [reply runtime/lua/man.lua]: It makes the semicolon optional, to match sequences like `\e[4;22m` as well as `\e[0m`

File: runtime/lua/man.lua [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+      end
+
+      overstrike = false
+      prev_char = ''
+      byte = byte + #char
+      chars[#chars + 1] = char
+    elseif escape then
+      -- Use prev_char to store the escape sequence
+      prev_char = prev_char .. char
+      local sgr = prev_char:match("^%[([\020-\063]*)m$")
```
[the remark this answers is not in the document]
Comment (keidax) [reply runtime/lua/man.lua]: Hmm, I forget why I chose that particular range, but `[\020-\063]` is wrong. For a valid CSI sequence, the bytes in between `[` and the final byte must be in the ASCII range 0x20-0x3f, so that should be `[\032-\063]`. I don't think it's mentioned in the xterm page, but this is part of the ECMA-48 standard, sect. 5.4. I'll add an explaining comment.

File: runtime/lua/man.lua [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+      if sgr then
+        local match = ''
+        while sgr and #sgr > 0 do
+          match, sgr = sgr:match("^(%d*);?(.*)")
+          add_attr_hl(match + 0) -- coerce to number
+        end
+        escape = false
+      elseif not prev_char:match("^%[[\020-\063]*$") then
```
[the remark this answers is not in the document]
Comment (keidax) [reply runtime/lua/man.lua]: Since it's iterating one character at a time, that's checking for a partial CSI sequence, and throwing away the sequence if it ends up being something else.

File: runtime/lua/man.lua (hunk 4 above)
Comment (keidax) [reply runtime/lua/man.lua]: This doesn't occur too often, but one example would be the man page for `mawk`. View it with `less`, search for `opt_expr`, and you'll see an underlined word with an underscore. The behavior of checking the previous overstrike and defaulting to bold is essentially what `less` does too.

File: runtime/lua/man.lua (hunk 6 above)
[the remark this answers is not in the document]
Comment (bfredl) [reply runtime/lua/man.lua]: As long as all Ctrl chars are ASCII, which I strongly suspect in man format, segmenting UTF-8 codepoints (which not unambiguously is "chars" anyway) shouldn't be needed. The highlight API only cares about bytes anyway.

File: test/functional/plugin/man_spec.lua [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+        b = { bold = true },
+        i = { italic = true },
+        u = { underline = true },
+        bi = { bold = true, italic = true },
+        biu = { bold = true, italic = true, underline = true },
+      },
+      {{ bold = true, foreground = Screen.colors.Blue }})
+    end
+
+    local function expect_without_highlights(string)
```
Comment (bfredl) [inline test/functional/plugin/man_spec.lua]: This is an anti-pattern. Just use `screen:snapshot_util()` and it will capture highlighting with not much extra work

File: test/functional/plugin/man_spec.lua [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+      command('syntax off') -- Ignore syntax groups
+      screen = Screen.new(52, 5)
+      screen:attach()
+    end)
+
+    after_each(function()
+      screen:detach()
+    end)
+
+    local function expect(string)
```
Comment (bfredl) [inline test/functional/plugin/man_spec.lua]: `screen:set_default_attr_ids`

File: test/functional/plugin/man_spec.lua [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+        biu = { bold = true, italic = true, underline = true },
+      },
+      {{ bold = true, foreground = Screen.colors.Blue }})
+    end
+
+    local function expect_without_highlights(string)
+      screen:expect(string, nil, true)
+    end
+
+    local function insert_lines(...)
```
Comment (bfredl) [inline test/functional/plugin/man_spec.lua]: `helpers.insert`

File: runtime/lua/man.lua (hunk 6 above)
[the remark this answers is not in the document]
Comment (keidax) [reply runtime/lua/man.lua]: Depending on locale settings, there could be a lot of overstruck text outside of ASCII range. Handling those multibyte text characters correctly requires splitting into codepoints.

[the remark this answers is not in the document]
Comment (bfredl) [reply runtime/lua/man.lua]: Is it well-defined somewhere that overstrike works on a codepoint at a time, and neither byte nor grapheme cluster (which is closer to user-perceived character)?

File: test/functional/plugin/man_spec.lua (hunk 13 above)
Comment (keidax) [reply test/functional/plugin/man_spec.lua]: I added this because the text contains a lot of control characters, and I found
```lua
insert_lines("i\bi")
```
more readable than
```lua
helpers.insert("i<C-v><C-h>i")
```
Would you prefer the second form?

Comment (justinmk) [reply test/functional/plugin/man_spec.lua]: @keidax The second form is preferred, it's the prevailing convention in our tests and throughout Vim-land.

File: runtime/lua/man.lua (hunk 6 above)
[the remark this answers is not in the document]
Comment (keidax) [reply runtime/lua/man.lua]: Based on my testing & digging in the `grotty` sources, it certainly looks that way.

**Comment by keidax** [comment]
Thanks for the feedback! I rebased and addressed all the comments.

File: test/functional/plugin/man_spec.lua [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+
+before_each(function()
+  plugin_helpers.reset()
+  helpers.clear()
+  command('syntax on')
+  command('set filetype=man')
+end)
+
+describe('In autoload/man.vim', function()
+  describe('function man#highlight_formatted_text', function()
```
Comment (justinmk) [inline test/functional/plugin/man_spec.lua]: Function name needs update

**Comment by justinmk** [comment]
LGTM thanks for reviews!

File: runtime/lua/man.lua (hunk 6 above)
[the remark this answers is not in the document]
Comment (nhooyr) [reply runtime/lua/man.lua]: Regardless of precedence, wouldn't it be more correct to handle overstrikes as applying to a grapheme and not a codepoint?

File: runtime/lua/man.lua (hunk 4 above)
Comment (nhooyr) [reply runtime/lua/man.lua]: Very odd, I'd expect that to be a space, followed by a backspace and then an underline, not a overstriken underline. Otherwise LGTM but do we need the `last_hl and last_hl[3] == byte` check?

**Review by nhooyr — COMMENTED** [review]
Made some comments on outdated issues.

1. https://github.com/neovim/neovim/pull/7623#discussion_r158958980
2. https://github.com/neovim/neovim/pull/7623#discussion_r158936051

**Comment by justinmk** [comment]
This works well, and performance seems good. Thanks everyone!

**Comment by aktau** [comment]
This doesn't work for me, I'm guessing because of `LUA_PATH` issues I was afraid of a while ago:

```
E5105: Error while calling lua chunk: [string "<VimL compiled string>"]:1: module 'man' not found
```

Still investigating.

**Comment by aktau** [comment]
On second thought, it may just be because the `man.lua` file is not installed:

```
11:20:01 ~/neovim » fname lua
./share/nvim/runtime/syntax/lua.vim
./share/nvim/runtime/doc/if_lua.txt
./share/nvim/runtime/indent/lua.vim
./share/nvim/runtime/ftplugin/lua.vim
```

**Comment by aktau** [comment]
Confirmed, the following makes it work:

```
11:25:38 neovim/runtime git:(master) » cp -r lua $HOME/neovim/share/nvim/runtime
```

(My install prefix is `-DCMAKE_INSTALL_PREFIX:PATH=$(HOME)/neovim`).
